### PR TITLE
Auto name RVs by parsing AST

### PR DESCRIPTION
**Pull request by twiecki**
Based on #83. See #79 for more discussion.

image

This code directly rewrite the bytecode of the function which is quite elegant because it all looks and runs the same. For example, debugging still works:
image

We can also easily detect cases where we can't infer the name and raise an exception that no name was provided.

### CODE REVIEW

**Comment by twiecki** [comment]
OK I think this is ready for review.

File: pymc4/tests/test_autoname.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+"""
+Tests for PyMC4 auto naming.
+"""
+
+from .. import model
+
+import pymc4 as pm
```
Comment (canyon289) [inline pymc4/tests/test_autoname.py]: Can these be changed to relative imports for consistency?

File: pymc4/tests/test_autoname.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+from pymc4 import Normal
+
+def create_rvs(**kwargs):
+    return pm.Normal(0, 1, name="test")
+
+
+@pm.model(auto_name=True)
```
Comment (canyon289) [inline pymc4/tests/test_autoname.py]: This could be placed in fixture to to keep global scope clean

File: pymc4/tests/test_autoname.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+def autoname_test():
+    mu = pm.Normal(0, 1)
+    mu2 = Normal(0, 1)
+    mu3 = pm.random_variables.Normal(0, 1)
+    pm.Normal(0, 1, name="custom")
+    dummy = pm.Normal(0, 1, name="custom2")
+    x = create_rvs()
+
+
+def test_auto_name():
```
Comment (canyon289) [inline pymc4/tests/test_autoname.py]: This test really shows off the how nice the API is!

File: pymc4/tests/test_autoname.py (hunk 2 above)
Comment (twiecki) [reply pymc4/tests/test_autoname.py]: Actually that doesn't work here as we can't parse AST nested functions. I added comments to explain that.

**Comment by ericmjl** [comment]
@twiecki ready to merge?

**Comment by twiecki** [comment]
yes!

**Comment by twiecki** [comment]
Ugh, this method doesn't seem to work with autograph which is run with `defun` :(.

**Comment by fonnesbeck** [comment]
Do you have an example of where this fails?

**Comment by twiecki** [comment]
@fonnesbeck e.g.
 
```python
@pm.model(auto_name=True)
def model():
    x = [pm.Normal(0, 1) for i in range(5)]
    pm.Normal(0, 1)
```

**Comment by twiecki** [comment]
> Ugh, this method doesn't seem to work with autograph which is run with defun :(.

I have a fix over here: https://github.com/pymc-devs/pymc4/pull/90/commits/3d55143c0da38e179075972f45ca025782bb717b it's not super clean but it does the trick.

**Comment by brandonwillard** [comment]
Is the `defun` issue solved by changing the decorator order?

**Comment by twiecki** [comment]
I don't think we can, as then we'd have to parse the super-mangled AST that `defun` leaves us. The current "hack" works quite well and I don't see a problem with it.

**Pull request merged**